_posts/2021-06-22-type-phantoms-and-safe-apis.md

```diff
@@ -0,0 +1,712 @@
+---
+layout: post
+title: Type Phantoms and Safe APIs
+author: mpajunen
+excerpt: >
+    Writing both frontend and backend of a web application in TypeScript allows using common model
+    type definitions in both. Let's explore one technique to make the API calls type-safe as well.
+tags:
+    - Api
+    - TypeScript
+---
+
+Using a typed language like TypeScript for the frontend of a web application can already be a great
+boon. When using a backend service also written in TypeScript, it's easy to share some model types.
+What about making the API boundary between the frontend and the backend type-safe? There are several
+ways to do that as well.
+
+## Approaches to API type safety
+
+One way to create a type-safe API is to generate code from metadata. If you have API documentation
+in OpenAPI or similar format, there are libraries that can generate corresponding types. Depending
+on the use case, some custom code might be needed as well. Tools like
+Apollo [rely on code generation](https://github.com/apollographql/apollo-tooling#code-generation)
+for type safety.
+
+An interesting approach is to first create runtime representations of the types in TypeScript. Types
+in TS are compile-time only but can be derived from the runtime code. This code can be used to
+validate data at the edges of the type system. [io-ts](https://gcanti.github.io/io-ts/) uses this
+approach.
+
+My colleague
+Jaakko [also blogged about typed APIs](https://dev.solita.fi/2018/09/06/fullstack-typed-api-with-typescript.html)
+earlier. Using type casts can enable some powerful features, such as runtime type checks.
+
+TypeScript has evolved a lot, and in this post we'll examine a types first approach. The main
+advantage is a clean API endpoint model, that can use all type system features. This model can be
+manipulated with type logic to create helpers and utilities as needed. It's not without downsides
+though: All validation logic will have to be written separately, since there are no runtime
+representations. Generating API documentation may require extra effort as well.
+
+Choosing the right approach depends on the problem: Is the API external, or only for internal use?
+How important is API documentation? What kind of validation logic is needed? Are you starting a
+greenfield project or expanding existing functionality?
+
+We'll start with an example that can be adapted to any typical REST API. This means using a normal
+route structure, some path parameters, and common HTTP methods. After a basic example, we'll explore
+expanding the functionality to cover real-world use cases and ponder a bit how all this could be
+simpler or even more useful.
+
+Complete source code, including a working example
+application, [is available on GitHub](https://github.com/mpajunen/rest-type-map).
```

Comment:
Linking this right at the start is a good call. 👌🏼

For extra points, how about plopping down a CodeSandbox for the example react-express app?

Reply:
Hmm... doesn't seem to be any easy way to set that up for separate client + server.

Reply:
Ah, good point! I hadn't considered that.